### ISSUE TITLE
Code and plugin contex questions

### ISSUE THREAD

**Issue by Michelvl92** [wang-xinyu/tensorrtx#903]
First of all, many thanks for making such a nice repo, with many nice implementations, as well as an explanation, starting with a simple mlp model up to more advanced models. This helped me a lot!

I am planning to implement Retinanet with resnet50 backbone, but first I need some extra understanding of some parts of the code, which are mainly missing in the simple examples.

1. in every `common.c` u use the on implemented `addBatchNorm2d()` and not Nvidia's batch norm, what is the reason for this, and why is such a "basic" operation a "tricky operation"?

2. e.g. in https://github.com/wang-xinyu/tensorrtx/blob/master/retinaface/retina_r50.cpp, and many other places and models used to do the following:
  ```
      float *deval = reinterpret_cast<float*>(malloc(sizeof(float) * 256 * 2 * 2));
      for (int i = 0; i < 256 * 2 * 2; i++) {
          deval[i] = 1.0;
      }
      Weights deconvwts{DataType::kFLOAT, deval, 256 * 2 * 2};
      IDeconvolutionLayer* up3 = network->addDeconvolutionNd(*output3->getOutput(0), 256, DimsHW{2, 2}, deconvwts, emptywts);
  ```
  Could you explain it in a little more detail?

3. The learning curve is going very steep when we switch to a "plugin". Calling and creating and adding a plugin is very clear (some context would be nice), but the definition of a plugin in the .h and .cu is a lot is going on! As I have medium experience in writing Cuda kernels, I am not asking for an explanation for the kernel, but everything around, how to start making an own plugin, and when should I know when I need such a plugin? It would be really nice having some context such as in the `mlp` example is done!

4. Is it possible with the Python API to add a plugin (the one in the  .h and .cu files) and add them to the model (which is defined with the Python API?)
5. Is it also possible to create a plugin with the Python API?
6. What is your experience in converting a model from onnx to TensorRT vs defining a model directly in TensorRT as in this repo in terms of speed?

**Comment by GeneralJing** [comment]
Is there no batchnorm layer in tensorrt about  question 1?

**Comment by wang-xinyu** [comment]
1. in every common.c u use the on implemented addBatchNorm2d() and not Nvidia's batch norm, what is the reason for this, and why is such a "basic" operation a "tricky operation"?

I didn't see Nvidia's batch norm.

**Comment by wang-xinyu** [comment]
2. e.g. in https://github.com/wang-xinyu/tensorrtx/blob/master/retinaface/retina_r50.cpp, and many other places and models used to do the following:

It is parsing weights for conv layer.

**Comment by wang-xinyu** [comment]
4 & 5.

We can use python api to add plugin layer,  but the plugin should be implemented in cuda and c++.
Just check the api documentation .

**Comment by wang-xinyu** [comment]
6.
For speed,  it's mostly the same,    won't have obvious difference.

**Comment by wang-xinyu** [comment]
3.
For plugin,  this is in higher level,   you can deepen into that after you get familiar with builtin layers.
